librz/il/vm_layer.c

```diff
@@ -0,0 +1,505 @@
+		cur_op_str = rz_str_newf("(%s %s)", rz_il_op2str(op->code), op->op.goto_->lbl);
+		helper[op->op.goto_->ret_ctrl_eff] = cur_op_str;
+		ret = op->op.goto_->ret_ctrl_eff;
+		break;
+		// 4 Int memebers
+	case RZIL_OP_INT:
+		cur_op_str = rz_str_newf("(%s %d)", rz_il_op2str(op->code), op->op.int_->value);
+		helper[op->op.int_->ret] = cur_op_str;
+		ret = op->op.int_->ret;
+		break;
+	case RZIL_OP_STORE:
+		arg1 = helper[op->op.store->key];
+		arg2 = helper[op->op.store->value];
+		cur_op_str = rz_str_newf("(%s %s %s)", rz_il_op2str(op->code), arg1, arg2);
+		helper[RZ_IL_VM_MAX_TEMP - 1] = cur_op_str; // op_store->ret == -1
+		ret = RZ_IL_VM_MAX_TEMP - 1;
+		break;
+	case RZIL_OP_LOAD:
+		arg1 = helper[op->op.load->key];
+		cur_op_str = rz_str_newf("(%s %s)", rz_il_op2str(op->code), arg1);
+		helper[op->op.load->ret] = cur_op_str;
+		ret = op->op.load->ret;
+		break;
+	case RZIL_OP_BRANCH:
+		// true or false may be an empty one,
+		// then true/false == -1
+		arg1 = helper[op->op.branch->condition];
+		arg2 = (op->op.branch->true_eff == -1) ? "<NOP>" : helper[op->op.branch->true_eff];
+		arg3 = (op->op.branch->false_eff == -1) ? "<NOP>" : helper[op->op.branch->false_eff];
+		cur_op_str = rz_str_newf("(%s %s %s %s)", rz_il_op2str(op->code), arg1, arg2, arg3);
+		helper[op->op.branch->ret] = cur_op_str;
+		ret = op->op.branch->ret;
+		break;
+	case RZIL_OP_UNK:
+	case RZIL_OP_B0:
+	case RZIL_OP_B1:
+		cur_op_str = rz_str_newf("%s", rz_il_op2str(op->code));
+		helper[op->op.unk->ret] = cur_op_str;
+		ret = op->op.unk->ret;
+		break;
+	case RZIL_OP_ITE:
+	case RZIL_OP_SHIFTR:
+	case RZIL_OP_SHIFTL:
+		arg1 = helper[op->op.ite->condition];
+		arg2 = helper[op->op.ite->x];
+		arg3 = helper[op->op.ite->y];
+		cur_op_str = rz_str_newf("(%s %s %s %s)", rz_il_op2str(op->code), arg1, arg2, arg3);
+		helper[op->op.ite->ret] = cur_op_str;
+		ret = op->op.ite->ret;
+		break;
+		// 3 Int members
+	case RZIL_OP_ADD:
+	case RZIL_OP_SUB:
+	case RZIL_OP_MUL:
+	case RZIL_OP_DIV:
+	case RZIL_OP_MOD:
+	case RZIL_OP_SDIV:
+	case RZIL_OP_SMOD:
+	case RZIL_OP_LOGXOR:
+	case RZIL_OP_LOGAND:
+	case RZIL_OP_LOGOR:
+	case RZIL_OP_ULE:
+	case RZIL_OP_SLE:
+	case RZIL_OP_SEQ:
+	case RZIL_OP_BLK:
+	case RZIL_OP_AND_:
+	case RZIL_OP_OR_:
+		arg1 = helper[op->op.add->x];
+		arg2 = helper[op->op.add->y];
+		cur_op_str = rz_str_newf("(%s %s %s)", rz_il_op2str(op->code), arg1, arg2);
+		helper[op->op.add->ret] = cur_op_str;
+		ret = op->op.add->ret;
+		break;
+	case RZIL_OP_MSB:
+	case RZIL_OP_LSB:
+	case RZIL_OP_NEG:
+	case RZIL_OP_NOT:
+	case RZIL_OP_JMP:
+	case RZIL_OP_INV:
+		arg1 = helper[op->op.inv->x];
+		cur_op_str = rz_str_newf("(%s %s)", rz_il_op2str(op->code), arg1);
+		helper[op->op.inv->ret] = cur_op_str;
+		ret = op->op.inv->ret;
+		break;
+	// Perform !! ---> Print !!
+	case RZIL_OP_PERFORM:
+		ret = op->op.perform->eff;
+		break;
+	default:
+		ret = 0;
+		printf("[WIP]\n");
+		break;
+	}
+
+	return ret;
+}
+
+void rz_il_vm_list_printer_step(RzPVector *op_list) {
```

Comment:
Move this function to `librz/core/cil.c`

Reply:
It looks unused btw, maybe could be removed?

Reply:
It's an experimental one. Used in the separated repo version. We can remove it now